Update ArcGisMapServerImageryProvider.js to add support for refreshing the token.

a-stacey opened the pull request:
I have left the following comment rather then addressing the issue because this is somewhat of an edge case (doesn't happen that often in practice) and it seems better to fire off multiple requests occasionally rather then complicate the code detecting and handling the code, but happy to resolve this issue if you feel this is better:
```
// Note: The token may have already been updated between the request and now (when the responce is recieved),
// but for now we don't detect and optomise for this case and send off a new token request regardless.
```

Here is some code that I mocked up to handle this properly:
```
+                let lastToken = that._token;
                 return loadImageViaBlob(url).otherwise(function(requestErrorEvent) {
                     // If the token has expired the server sets the HTTP status code to 498 specifically to indicate this error.
-                    if ((requestErrorEvent.statusCode === 498) && (tokenRetries > 0)) {
-                        tokenRetries--;
-
-                        // Note: The token may have already been updated between the request and now (when the responce is recieved),
-                        // but for now we don't detect and optomise for this case and send off a new token request regardless.
-                        return updateToken(that).then(() => {
-                            // Rebuild the URL now that the token has been updated.
-                            url = buildImageUrl(that, x, y, level);
-                            return loadImageWithToken();
-                        });
+                    if (requestErrorEvent.statusCode === 498) {
+
+                    // If the token has already been updated.
+                    if (lastToken !== that._token) {
+                        return loadJsonHandleError();
+                    } else {
+                        if (tokenRetries > 0) {
+                            tokenRetries--;
+                            return updateToken(that).then(() => {
+                                // Rebuild the URL now that the token has been updated.
+                                url = buildImageUrl(that, x, y, level);
+                                return loadImageWithToken();
+                            });
+                        }
                     }
+                }
```

The reason that I don't particularly like this is the intuition jump that you need from `lastToken = that._token` being a proxy for the value of the `url` component of `loadImageViaBlob(url)` changing which I think inhibits the maintainability of the code. Technically you could check if `url` has changed, but this is not particularly clean either.

Commit:
[TerriaJS/terriaJS#2588] ArcGisMapServerImagryProvider: Remove stray whitespace from file before editing further.

Commit:
[TerriaJS/terriaJS#2588] ArcGisMapServerImageryProvider: Refactor the URL creation out of pickFeatures() into a separate function.

Commit:
[TerriaJS/terriaJS#2588] ArcGisMapServerImageryProvider: Refactor the feature creation out of pickFeatures() into a separate function.

Commit:
[TerriaJS/terriaJS#2588] ArcGisMapServerImageryProvider: Add a .token setter to allow updating of tokens.

Commit:
[TerriaJS/terriaJS#2588] ArcGisMapServerImageryProvider: Add a callback argument to allow for more tokens to be requested if they expire.

Commit:
[TerriaJS/terriaJS#2588] ArcGisMapServerImageryProvider: Add ._requestNewToken() call to requestImage() and pickFeatures() to refresh the token when it expires.

Commit:
[TerriaJS/terriaJS#2588] ArcGisMapServerImageryProvider: Partially optomise the token request to reduce multiple requests.

We optomise so that if a token request is already in progress we notify all requesters rather then sending off a new request for each attempt.

Commit:
[TerriaJS/terriaJS#2588] ArcGisMapServerImageryProvider: Refactor to add a function that explicitly updates the token rather then repeating the operation.

Commit:
[TerriaJS/terriaJS#2588] ArcGisMapServerImageryProvider: Add 499 error code to set of errors to update the token for.

499 Indicates that no token was specified when one was required.

kring commented on Source/Scene/ArcGisMapServerImageryProvider.js:
```diff
@@ -59,6 +61,7 @@ define([
      * @param {Object} options Object with the following properties:
      * @param {String} options.url The URL of the ArcGIS MapServer service.
      * @param {String} [options.token] The ArcGIS token used to authenticate with the ArcGIS MapServer service.
+     * @param {Promise.<String>} [options.requestNewToken] A callback to retrieve new tokens if its detected that the current token has expired.
```
`requestNewToken` is a function, not a `Promise<String>`.  Here's how you specify that a parameter is a callback function with jsdoc: http://usejsdoc.org/tags-callback.html

a-stacey replied:
Thanks for this.

26801a9ca40e64314f582ac743c567c13c7de5e5

kring commented on Source/Scene/ArcGisMapServerImageryProvider.js:
```diff
@@ -311,14 +400,20 @@ define([
+
+                if (defined(token)) {
+                    this._token = token;
+                }
```
I can't think of any reason to delete the token and then only add it if it's defined.  A simple `this._token = token` will do, and will make `_token` undefined if the new token is undefined.  There are very few places where it's a good idea to use delete.

a-stacey replied:
>  There are very few places where it's a good idea to use delete.

Is there any reason behind this?

Have changed this.

This design was put in place so as not to pollute this._token (i.e. so that the class state would not have ._token defined if it had been defined and then undefined exactly the same as when it was never defined). Can't imagine any practical differences between the two implementation, but would expect `for (var key in obj) { console.log(key); }` to differ for the two cases.

kring replied:
Why do you want `_token` to not exist when it's not used?  In a statically-typed language (e.g. Java), you'd declare a field for the token and you wouldn't try to change the type depending on whether the field was used or not, right?  But that's essentially what you're doing by deleting the field.  You're forcing the browser to re-allocate the object without the field, or perhaps drop the whole object into (slower) hash mode.  Admittedly performance is not much of a concern here, but there's no reason to make the browser do extra work when `_token` should just be a declared field on the object like all the other fields.

kring replied:
Matt and I described this in more detail in our chapter in WebGL Insights.  Check out section 4.3.1 here: http://webglinsights.github.io/downloads/WebGL-Insights-Chapter-4.pdf

kring commented on Source/Scene/ArcGisMapServerImageryProvider.js:
```diff
@@ -605,7 +700,32 @@ define([
+        if (!defined(this._requestNewToken)) {
+            return ImageryProvider.loadImage(this, url);
+        } else {
+            let that = this;
```
You need to stick to ES5 in Cesium.  `var` instead of `let`.

a-stacey replied:
Thanks.

06470fe74d45a3578ba6b872456df4f1d668fba3

kring commented on Source/Scene/ArcGisMapServerImageryProvider.js:
```diff
@@ -605,7 +700,32 @@ define([
+                    if (((requestErrorEvent.statusCode === 498) || (requestErrorEvent.statusCode === 499)) && (tokenRetries > 0)) {
+                        tokenRetries--;
+
+                        // Note: The token may have already been updated between the request and now (when the responce is recieved),
```
typo: recieved -> received

kring replied:
Also responce -> response

a-stacey replied:
Thanks.

58a010d5f3709f636030ac9429a5b2ddc59bcade

kring commented on Source/Scene/ArcGisMapServerImageryProvider.js:
```diff
@@ -605,7 +700,32 @@ define([
+                        tokenRetries--;
+
+                        // Note: The token may have already been updated between the request and now (when the responce is recieved),
+                        // but for now we don't detect and optomise for this case and send off a new token request regardless.
```
optomise -> optimize

a-stacey replied:
Thanks.

58a010d5f3709f636030ac9429a5b2ddc59bcade

kring commented on Source/Scene/ArcGisMapServerImageryProvider.js:
```diff
@@ -297,6 +301,91 @@ define([
         return url;
     }
 
+    function buildPickURL(imageryProvider, x, y, level, longitude, latitude) {
```
Move these private functions to the bottom, after all the class stuff.

a-stacey replied:
Thanks, Have also move the existing private function buildImageUrl()

7aca562e32f4a173d1a418e909b7580703245a93

kring commented on Source/Scene/ArcGisMapServerImageryProvider.js:
```diff
@@ -297,6 +301,91 @@ define([
+    function updateToken(imageryProvider) {
+        if (!defined(imageryProvider._newTokenRequestInFlight) && defined(imageryProvider._requestNewToken)) {
+            imageryProvider._newTokenRequestInFlight = imageryProvider._requestNewToken().then(function(newToken) {
+                delete imageryProvider._newTokenRequestInFlight;
```
Set it undefined, don't delete it.

a-stacey replied:
No worries...but why? just trying to learn why this is bad style here.

Thanks.

kring commented on Source/Scene/ArcGisMapServerImageryProvider.js:
```diff
@@ -297,6 +301,91 @@ define([
+                delete imageryProvider._newTokenRequestInFlight;
+                imageryProvider.token = newToken;
+                return newToken;
+            });
```
Super subtle bug possibility here.  What happens if `imageryProvider._requestNewToken()` returns a promise that is already resolved?  In the version of when.js we're currently using, the function passed to `.then` will be executed immediately, before the call to `then()` even returns.  Which means `_newTokenRequestInFlight` will be set to undefined before it even gets assigned the promise in the first place!  And then when `then()` returns, the (already resolved) promise it returns will be assigned to `imageryProvider._newTokenRequestInFlight`.  End result: we think there's a request in flight when there's actually not.

kring replied:
You can fix this using an approach like this:
```
var loadPromise = imageryProvider._requestNewToken().then(function(newToken) {
    imageryProvider.token = newToken;
    return newToken;
});

var resolvedOrRejected = false;
loadPromise.always(function() {
    resolvedOrRejected = true;
    imageryProvider._newTokenRequestInFlight = undefined;
});

imageryProvider._newTokenRequestInFlight = resolvedOrRejected ? undefined : loadPromise;
```

Yes, this is kind of terrible, and is one reason that all promise implementations have moved to not immediately calling the callback.

a-stacey replied:
Thanks.

8c079608751246ac1a0870cdc2e4e208aca8a5dd

kring commented:
That's all I have!

kring commented on Source/Scene/ArcGisMapServerImageryProvider.js:
```diff
@@ -605,7 +700,32 @@ define([
+            let that = this;
+            let tokenRetries = 1;
+            function loadImageWithToken () {
+                return loadImageViaBlob(url).otherwise(function(requestErrorEvent) {
```
Oops one more: you need the `throttleRequestByServer` call.  See `ImageryProvider.loadImage`.

a-stacey replied:
Ah right.

ae96af9afe12d1934405375d1a97d0fe5d642f09

Commit:
[TerriaJS/terriaJS#2588] ArcGisMapServerImageryProvider: Change let's to vars to make ES5 compliant.

This issue was identified during code review.

Commit:
[TerriaJS/terriaJS#2588] ArcGisMapServerImageryProvider: Fix spelling errors.

These issues were identified during code review.

Commit:
[TerriaJS/terriaJS#2588] ArcGisMapServerImageryProvider: Simplify token setter.

This issue was identified during code review.

Commit:
[TerriaJS/terriaJS#2588] ArcGisMapServerImageryProvider: Change from delete to undefined when imageryProvider._newTokenRequestInFlight is resolved.

This issue was identified during code review.

Commit:
[TerriaJS/terriaJS#2588] ArcGisMapServerImageryProvider: Throttle the image requests when tokens are used in .requestImage().

This issue was identified during code review.

Commit:
[TerriaJS/terriaJS#2588] ArcGisMapServerImageryProvider: Fix a subtle bug in the order of ._newTokenRequestInFlight definition.

This issue was identified during code review.

The issue identified by @kring was:
> Super subtle bug possibility here. What happens if imageryProvider._requestNewToken() returns a promise that is already resolved? In the version of when.js we're currently using, the function passed to .then will be executed immediately, before the call to then() even returns. Which means _newTokenRequestInFlight will be set to undefined before it even gets assigned the promise in the first place! And then when then() returns, the (already resolved) promise it returns will be assigned to imageryProvider._newTokenRequestInFlight. End result: we think there's a request in flight when there's actually not.

Commit:
[TerriaJS/terriaJS#2588] Add change log entry for this branch.

Commit:
[TerriaJS/terriaJS#2588] ArcGisMapServerImageryProvider: Fix documentation of requestNewToken callback.

    This issue was identified during code review.

Commit:
[TerriaJS/terriaJS#2588] ArcGisMapServerImageryProvider: Relocated private functions to the bottom of the file.

This relocation includes the existing function: buildImageUrl().

This issue was identified during code review.

Commit:
[TerriaJS/terriaJS#2588] ArcGisMapServerImageryProvider: Refactor throttleRequestByServer() call.

Now it is only called a single time per image. Previously it was being called internally from within loadImageWithToken(url) when the token had expired, which meant that when it became undefined (quite common) it would lead to the layer failing.

a-stacey commented:
Please let me know what you think of `url` in ` function loadImageWithToken (url) {` which masks `var url = buildImageUrl(this, x, y, level);`.  They do happen to be the same thing, but for maintenance I could see this leading to subtle errors being introduced. I feel that its not exactly ideal and could be a bit confusing (readability)  but no other name particularly makes sense.

In: d8bd6017dfb021c3468e1c2e8359995e6ad2bd9f

Commit:
[TerriaJS/terriaJS#2588] ArcGisMapServerImageryProvider: Refactor requestMetadata() to use .then() .otherwise() rather then when().

Commit:
[TerriaJS/terriaJS#2588] ArcGisMapServerImageryProvider: Refactor requestMetadata() to call an intermediate function ahead of adding code to update the token.

Commit:
[TerriaJS/terriaJS#2588] ArcGisMapServerImageryProvider: Handle if the token has expired from requestMetadata().

Commit:
[TerriaJS/terriaJS#2588] ArcGisMapServerImageryProvider: Refactor handling token errors with JSON responses in pickFeatures().

Commit:
[TerriaJS/terriaJS#2588] ArcGisMapServerImageryProvider: Refactor handling token errors with JSON responses more generic.

Commit:
[TerriaJS/terriaJS#2588] ArcGisMapServerImageryProvider: Use the refactored JSON handling code from requestMetadata() to remove code duplication.

Commit:
[TerriaJS/terriaJS#2588] ArcGisMapServerImageryProvider: Fix lint warnings.

Commit:
[TerriaJS/terriaJS#2588] ArcGisMapServerImageryProvider: Move function declaration out of else to remove lint warning.

a-stacey commented:
One additional thing to watch out for, I'm not sure if I handle the case correctly where we try getting a valid token (once), but its still not valid one the second attempt. The sentential prevents us from entering an infinite loop, but we just terminate...allowing other handlers to deal with the problem as if there were no token error handling code. See:
```
                if (((requestErrorEvent.statusCode === 498) || (requestErrorEvent.statusCode === 499)) && (tokenRetries > 0)) {
and
                    if (((json.error.code === 498) || (json.error.code === 499)) && defined(item._requestNewToken) && (tokenRetries > 0)) {
```

In practice hopefully this case is extremely unlikely, but not sure if what I have done here is fine, or whether I should explicitly throw an error.

kring committed:
Remove redundant section
